src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp

```diff
@@ -1143,12 +1163,13 @@ MarkRanges MergeTreeDataSelectExecutor::markRangesFromPKRange(
     std::vector<FieldRef> part_offset_left(2);
     std::vector<FieldRef> part_offset_right(2);
 
-    auto may_be_true_in_range = [&](MarkRange & range)
+    auto check_in_range = [&](const MarkRange & range, bool check_can_be_false = false)
     {
-        bool key_condition_maybe_true = true;
-        if (!key_condition.alwaysUnknownOrTrue())
+        BoolMask initial_mask = check_can_be_false ? BoolMask() : BoolMask::consider_only_can_be_true;
+
+        auto check_key_condition = [&]()
         {
-            if (range.end == marks_count && !has_final_mark)
```

Comment:
why?

Reply:
It's to ease the implementation of finding exact ranges (end points are important). Will refactor this in separate PR.